Added alpha overloads for add/sub ops with lists

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43926 Rewrote Adadelta with _foreach APIs
* #43925 Rewrote adamax optimizer with _foreach APIs
* #43727 Rewrote asgd optimizer with *_foreach APIs
* #43726 Rewrote rprop optimizer with _foreach APIs
* #43725 Rewrote rmsprop with _forerach APIs
* #43724 Rewrote SGD with _foreach APIs
* #43562 Rewrote AdamW with _foreach APIs
* #43507 Rewrote adam optimizer with foreach APIs
* **#43413 Added alpha overloads for add/sub ops with lists**

Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same dtype.
- All Tensors must be on the same device. 


----------------
**In this PR**
Added alpha overloads for add/sub ops with lists

**Tests**
Tested via unit tests

Commit:
WIP. Rewriting Adam with foreach

[ghstack-poisoned]

Commit:
Update on "WIP. Rewriting Adam with foreach"




[ghstack-poisoned]

izdeby commented on test/test_foreach.py:
```diff
@@ -19,13 +19,17 @@ class TestForeach(TestCase):
+        self.assertEqual([torch.sqrt(torch.ones(2, 2, device=device, dtype=dtype)) for _ in range(2)], res)
         self.assertEqual(tensors, res)
+        self.assertEqual(exp, res)
+        print(exp, res)
```
del

Commit:
Update on "WIP. Added alpha overloads for add/sub ops with lists"




[ghstack-poisoned]

Commit:
Update on "WIP. Added alpha overloads for add/sub ops with lists"




[ghstack-poisoned]

Commit:
Update on "WIP. Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "WIP. Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "WIP. Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "WIP. Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "WIP. Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "WIP. Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

cpuhrsch commented on test/test_foreach.py:
```diff
@@ -37,6 +37,19 @@ def _test_bin_op_list(self, device, dtype, foreach_op, foreach_op_, torch_op, N=
         self.assertEqual(res, tensors1)
         self.assertEqual(tensors1, expected)
 
+    def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_op, N=20):
```
Could add parameterize on binary function

Commit:
Update on "Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

[ghstack-poisoned]

Commit:
Update on "Added alpha overloads for add/sub ops with lists"


Differential Revision: [D23331896](https://our.internmc.facebook.com/intern/diff/D23331896)

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same dtype.
- All Tensors must be on the same device. 


----------------
**In this PR**
Added alpha overloads for add/sub ops with lists

**Tests**
Tested via unit tests


[ghstack-poisoned]

cpuhrsch reviewed (approved):
LGTM
